File: Python3/프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque

def solution(land):
    answer = 0
    direction = [(0,-1), (0,1), (-1,0), (1,0)]
    n = len(land)
    m = len(land[0])
    
    def bfs(start_i, start_j, index):
        que = deque([(start_i, start_j)])
        
        storage = 1
        while que:
            i, j = que.popleft()
            
            for di, dj in direction:
                ni,nj = di+i,dj+j
                if 0 <= ni < n and 0 <= nj < m and not visited[ni][nj] and land[ni][nj]:
                    visited[ni][nj] = index
                    storage += 1        
                    que.append((ni,nj))
        
        return storage 
    
    
    
    visited = [[0] * m for _ in range(n)]
    cnt = 1
    oils = [0]
    
    for i in range(n):
        for j in range(m):
            if not visited[i][j] and land[i][j]:
                visited[i][j]= cnt
                oils.append(bfs(i, j, cnt))
                cnt += 1
    
    print(oils)
    
    # [print(visited[a]) for a in range(n)]
    v = set()
    for j in range(m):
        v.clear()
        oil = 0
        for i in range(n):
            if visited[i][j] not in v:
                v.add(visited[i][j])
                oil += oils[visited[i][j]]
        # print(oil, v)
        answer = max(answer, oil)        
    
    return answer

File: JavaScript/프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
function solution(land) {
    let answer = 0;
    
    const [n, m] = [land.length, land[0].length];
    const visited = [];
    for (let i=0; i< n; i++) {
        visited.push(Array(m).fill(0));
    }
    
    const dir = [[0,-1], [0,1], [-1,0], [1,0]];
    
    const bfs = (startI, startJ, index) => {
        const que = {};
        let [front, rear] = [0,0];
        let storage = 1;
        que[rear++] = [startI, startJ];
        while (front !== rear) {
            const [i,j] = que[front++];
            for (let [di, dj] of dir){
                const [ni, nj] = [i+di, j+dj];
                if (0 <= ni && ni < n && 0 <= nj && nj < m && visited[ni][nj] === 0 && land[ni][nj] === 1) {
                    que[rear++] = [ni,nj];
                    visited[ni][nj] = index;
                    storage ++;
                }
            }
        }
        return storage;
    }
    
    let cnt = 1;
    const oils = [0];
    
    
    for (let i=0; i< n; i++){
        for (let j=0; j< m; j++){
            if (visited[i][j] === 0 && land[i][j] === 1) {
                visited[i][j] = cnt
                oils.push(bfs(i,j,cnt++))
            }
        }
    }
    
    
    const v = new Set();
    
    for (let j=0; j<m; j++){
        let oil = 0;
        v.clear();
        for (let i=0; i<n; i++){
            if (!v.has(visited[i][j])){
                v.add(visited[i][j]);
                oil += oils[visited[i][j]];
            }
        }
        answer = Math.max(answer, oil)
    }
    
    return answer;
}